### PR TITLE
Handle SIGTERM better in cluster agent

### PR DESCRIPTION
**Pull request by dharmit**
## Issue: 

#43020

## Problem

Presently, when cattle-cluster-agent receives a SIGTERM, it exits with non-zero exit code instead of handling it cleanly.

## Solution

The solution proposed in this PR continues the `logrus.Fatalf` behaviour if context returns an error other than `context.Canceled`.
 
## Testing

As mentioned in #43020 

### Manual Testing

1. Built images with proposed code change using: `TAG="dev-tag" REPO=<dockerhub-username> make quick`
1. Started rancher using rancher & rancher-agent images built using the changes proposed in this PR.
1. Import the existing cluster in Rancher UI and execute the curl command presented on the UI:
    ```sh
    $ curl --insecure -sfL https://192.168.29.16/v3/import/9lcgghmxz4fwtnmwqxrjm4czmz7hfglqx7zc4b25mckkxg4h62hj69_c-m-cjqp2626.yaml | kubectl apply -f
    ```

### Automated Testing

* Test types added/modified:
    * None
* If "None" - Reason: Lack of the framework capable of testing this fix/change
  
* If "None" - GH Issue/PR: https://github.com/rancher/rancher/issues/48105

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: pkg/agent/rancher/rancher.go [hunk 1]
```diff
@@ -87,8 +88,12 @@ func (h *handler) startRancher() {
 		logrus.Fatalf("Embedded rancher failed to initialize: %v", err)
 	}
 	go func() {
-		err = server.ListenAndServe(h.ctx)
-		logrus.Fatalf("Embedded rancher failed to start: %v", err)
+		if err := server.ListenAndServe(h.ctx); errors.Is(err, context.Canceled) {
+			// usually (not always) context cancellation would happen when cancel() corresponding to h.ctx gets called
```
Comment (snasovich) [inline pkg/agent/rancher/rancher.go:92]: "usually (not always)" is not very convincing - sounds like we're not 100% sure it's OK to ignore error in this case. Could you please clarify what other situation may happen that prompted to "not always" comment to be added?

Comment (dharmit) [reply pkg/agent/rancher/rancher.go:92]: > Could you please clarify what other situation may happen that prompted to "not always" comment to be added?

TBH, it was our discussion on Slack that prompted me to add this. AFAICS, the `h.ctx` here comes from: https://github.com/rancher/rancher/blob/c7dc9ae7ddca5a0847233a3c870a2946fb2fc8ae/cmd/agent/main.go#L184 https://github.com/rancher/rancher/blob/c7dc9ae7ddca5a0847233a3c870a2946fb2fc8ae/cmd/agent/main.go#L352-L358

Between that and below, I don't see the `topContext` being modified to become a child context of another context: https://github.com/rancher/rancher/blob/c7dc9ae7ddca5a0847233a3c870a2946fb2fc8ae/pkg/rancher/rancher.go#L330-L331

So, in this particular case, I don't know if there'll be any situation except receiving the signals `os.SIGINT` or `syscall.SIGTERM` where we'd encounter `context.Canceled`.

Correct me if I'm mistaken.

Comment (dharmit) [reply pkg/agent/rancher/rancher.go]: Btw, I figured a bug in the PR. Have addressed it and also modified the comment. Let me know if that makes sense.

File: pkg/agent/rancher/rancher.go [hunk 2]
```diff
@@ -88,7 +89,16 @@ func (h *handler) startRancher() {
 	}
 	go func() {
 		err = server.ListenAndServe(h.ctx)
-		logrus.Fatalf("Embedded rancher failed to start: %v", err)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
```
[the remark this answers is not in the document]
Comment (dharmit) [reply pkg/agent/rancher/rancher.go:93]: It's already happening in underlying wrangler code that's being used to set up the signal context: https://github.com/rancher/rancher/blob/c7dc9ae7ddca5a0847233a3c870a2946fb2fc8ae/cmd/agent/main.go#L184

So when one does `kill -s TERM 1` inside the cluster agent Pod, it prints:
```
time="2024-10-14T15:07:17Z" level=info msg="signal received: \"terminated\", canceling context..."
```

**Review by snasovich — APPROVED** [review]

**Pull request merged**